### PR TITLE
[Merged by Bors] - feat: iteratedDeriv_const_{s,}mul, iteratedDeriv_{c,}exp_const_mul

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
Based on @CBirkbeck's work on modular forms.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ADedecker** [comment]
The only part that is specific about `exp` is that `iteratedDeriv n exp = exp` right? If so could you prove the general result instead, and then use [iteratedDeriv_eq_iterate](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/IteratedDeriv.html#iteratedDeriv_eq_iterate) and [Real.iter_deriv_exp](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/SpecialFunctions/ExpDeriv.html#Real.iter_deriv_exp) to get your result?

**Review by j-loreaux — COMMENTED** [review]
~~Is there any reason not to do this for `NormedSpace.exp` directly? (of course, we'll still want the versions specialized to `Complex.exp` and `Real.exp` too.)~~ Anatole's suggestion is far better.

**Comment by ADedecker** [comment]
> Is there any reason not to do this for `NormedSpace.exp` directly? (of course, we'll still want the versions specialized to `Complex.exp` and `Real.exp` too.)

I was asking to ask for that, but then realized that we could refactor this whole file to go through `NormedSpace.exp`, so maybe we can do that in one go later. Also, I think the real generalization is that this is not a lemma about exponentials at all (although the fact that I'm the only one to notice it makes me worry about my own sanity) : `iteratedDeriv n (f (c * .)) x = c ^ n * iteratedDeriv n f (c * x)` is true for all `C^n` functions, right ?

**Comment by ADedecker** [comment]
Also, doing the proof in the general setting for `NormedSpace.exp` might be unexpectedly tricky, `iteratedFDeriv` is an annoying beast to fight...

**Comment by Ruben-VandeVelde** [comment]
Seems true. I'll take a look at updating the PR later
```lean
import Mathlib

variable {𝕜 : Type*} [NontriviallyNormedField 𝕜] [NormedAlgebra 𝕜 ℂ]

theorem xxx (n : ℕ) (c : 𝕜) (f : 𝕜 → 𝕜) (h : ContDiff 𝕜 n f) : iteratedDeriv n (fun x => f (c * x)) = fun x => c ^ n * iteratedDeriv n f (c * x) := by
  induction n with
  | zero => simp
  | succ n ih =>
    funext x
    rw [pow_succ, iteratedDeriv_succ, ih h.of_succ, deriv_const_mul_field]
    rw [iteratedDeriv_succ]
    rw [← Function.comp_def ]
    rw [deriv.comp x (h.differentiable_iteratedDeriv _ (Nat.cast_lt.mpr n.lt_succ_self)).differentiableAt (differentiableAt_id'.const_mul _)]
    rw [deriv_const_mul _ differentiableAt_id', deriv_id'']
    ring

theorem iteratedDeriv_cexp_const_mul_ (n : ℕ) (c : ℂ) :
    (iteratedDeriv n fun s : ℂ => Complex.exp (c * s)) = fun s => c ^ n * Complex.exp (c * s) := by
  rw [xxx]
  funext x
  rw [@iteratedDeriv_eq_iterate]
  rw [Complex.iter_deriv_exp]
```

**Comment by j-loreaux** [comment]
@Ruben-VandeVelde this still has a build failure

**Comment by j-loreaux** [comment]
I think this should go in `Analysis.Calculus.IteratedDeriv.Lemmas` (I know there has been some file shuffling recently).

File: Mathlib/Analysis/SpecialFunctions/ExpDeriv.lean [hunk 1]
```diff
@@ -174,6 +177,14 @@ theorem ContDiffWithinAt.cexp {n} (hf : ContDiffWithinAt 𝕜 n f s x) :
 
 end
 
+open Complex in
+theorem iteratedDeriv_cexp_const_mul (n : ℕ) (c : ℂ) :
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/ExpDeriv.lean:181]: I think the only real reason to have it (separate from the more general one) is if we mark it `simp`.
```suggestion
@[simp]
theorem iteratedDeriv_cexp_const_mul (n : ℕ) (c : ℂ) :
```

File: Mathlib/Analysis/SpecialFunctions/ExpDeriv.lean [hunk 2]
```diff
@@ -319,3 +330,8 @@ theorem fderiv_exp (hc : DifferentiableAt ℝ f x) :
 #align fderiv_exp fderiv_exp
 
 end
+
+open Real in
+theorem iteratedDeriv_exp_const_mul (n : ℕ) (c : ℝ) :
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/ExpDeriv.lean:335]:
```suggestion
@[simp]
theorem iteratedDeriv_exp_const_mul (n : ℕ) (c : ℝ) :
```

**Review by j-loreaux — APPROVED** [review]
Thanks, I think these should be `@[simp]`.

bors d+

**Comment by Ruben-VandeVelde** [comment]
bors merge

**Pull request closed without merging**